util/rlp/src/stream.rs

```diff
@@ -28,32 +29,75 @@ impl ListInfo {
 	}
 }
 
-/// Appendable rlp encoder.
-pub struct RlpStream {
-	unfinished_lists: ElasticArray16<ListInfo>,
-	buffer: ElasticArray1024<u8>,
-	finished_list: bool,
+pub trait RlpBuffer: Default + ops::Deref<Target = [u8]> + ops::DerefMut {
+	fn push(&mut self, e: u8);
+
+	fn append_slice(&mut self, elements: &[u8]);
+
+	fn clear(&mut self);
+
+	fn into_vec(self) -> Vec<u8>;
+
+	fn insert_slice(&mut self, index: usize, elements: &[u8]);
 }
 
-impl Default for RlpStream {
-	fn default() -> Self {
-		RlpStream::new()
+macro_rules! impl_rlp_buffer_for_elastic_array {
+	($name: ident) => {
+		impl RlpBuffer for $name<u8> {
+			#[inline]
+			fn push(&mut self, e: u8) {
+				self.push(e)
+			}
+
+			#[inline]
+			fn append_slice(&mut self, elements: &[u8]) {
+				self.append_slice(elements)
+			}
+
+			#[inline]
+			fn clear(&mut self) {
+				self.clear()
+			}
+
+			#[inline]
+			fn into_vec(self) -> Vec<u8> {
+				self.into_vec()
+			}
+
+			#[inline]
+			fn insert_slice(&mut self, index: usize, elements: &[u8]) {
+				self.insert_slice(index, elements)
+			}
+		}
 	}
 }
 
-impl RlpStream {
+impl_rlp_buffer_for_elastic_array!(ElasticArray8);
+impl_rlp_buffer_for_elastic_array!(ElasticArray1024);
+
+/// Should be used to encode should stream of rlp items.
+pub type RlpShortStream  = RlpConfigurableStream<ElasticArray8<u8>>;
+
+/// Should be used to encode stream of rlp items.
+pub type RlpStream = RlpConfigurableStream<ElasticArray1024<u8>>;
```

Comment:
As suggested in other comment, would extract `RlpStream` to a trait and then implement that for every kind of `RlpConfigurableStream` to simplify method signatures.